graph_weather/models/layers/thermalizer.py

```diff
@@ -0,0 +1,290 @@
+            nn.Conv2d(out_channels, out_channels, kernel_size=kernel_size, padding=padding),
+            nn.GroupNorm(min(8, out_channels), out_channels),
+            nn.ReLU(),
+            nn.MaxPool2d(kernel_size=3, stride=2, padding=1),
+        )
+        return contract
+
+    def _expand_block(
+        self, in_channels: int, out_channels: int, kernel_size: int, padding: int
+    ) -> nn.Sequential:
+        """Create an expanding block for the UNet."""
+        expand = nn.Sequential(
+            nn.Conv2d(in_channels, out_channels, kernel_size, padding=padding),
+            nn.GroupNorm(min(8, out_channels), out_channels),
+            nn.ReLU(),
+            nn.Conv2d(out_channels, out_channels, kernel_size, padding=padding),
+            nn.GroupNorm(min(8, out_channels), out_channels),
+            nn.ReLU(),
+            nn.ConvTranspose2d(
+                out_channels,
+                out_channels,
+                kernel_size=3,
+                stride=2,
+                padding=1,
+                output_padding=1,
+            ),
+        )
+        return expand
+
+
+class ThermalizerLayer(nn.Module):
+    """Thermalizer layer for inference-time denoising using diffusion models."""
+
+    def __init__(self, input_dim: int = 256, timesteps: int = 1000) -> None:
+        """Initialize the Thermalizer layer.
+
+        Args:
+            input_dim: Dimension of input features
+            timesteps: Number of diffusion timesteps
+        """
+        super().__init__()
+        self.score_model = AdaptiveUNet(input_dim, input_dim)
+        self.timesteps = timesteps
+        self.betas = self._cosine_beta_schedule(timesteps)
+        self.alphas = 1.0 - self.betas
+        self.alphas_cumprod = torch.cumprod(self.alphas, axis=0)
+
+    def forward(
+        self,
+        x: torch.Tensor,
+        t: torch.Tensor,
+        height: int = None,
+        width: int = None,
+        batch: int = None,
+    ) -> torch.Tensor:
+        """Apply the Thermalizer denoising step.
+
+        Args:
+            x: Input tensor of shape (batch * nodes, features)
+            t: Timestep (can be an integer scalar)
+            height: Optional height dimension (will be inferred if not provided)
+            width: Optional width dimension (will be inferred if not provided)
+            batch: Optional batch size (will be inferred if not provided)
+
+        Returns:
+            The denoised tensor of same shape as input
+        """
+        total_nodes, features = x.shape
+
+        # If dimensions not provided, try to infer them
+        if height is None or width is None or batch is None:
+            if batch is None:
+                # Assume batch size of 1 if not specified
+                batch = 1
+                nodes = total_nodes
+            else:
+                nodes = total_nodes // batch
+
+            if height is None or width is None:
+                height, width = self._infer_grid_dimensions(nodes)
```

Comment:
For this, just so I have a better understanding, does it matter if the input graph is arbitrary number and placement of nodes? Or does it have to be equally spaced nodes? My thought would be for a model that has a high resolution area with a lot more nodes, and then a lower resolution global grid, would that work here?